sycl/source/detail/async_alloc.cpp

```diff
@@ -46,21 +46,36 @@ void *async_malloc(sycl::handler &h, sycl::usm::alloc kind, size_t size) {
 
   auto &Adapter = h.getContextImplPtr()->getAdapter();
 
-  // Get events to wait on.
-  auto depEvents = getUrEvents(h.impl->CGData.MEvents);
-  uint32_t numEvents = h.impl->CGData.MEvents.size();
+  // Get CG event dependencies for this allocation.
+  const auto &DepEvents = h.impl->CGData.MEvents;
+  auto UREvents = getUrEvents(DepEvents);
+  uint32_t numEvents = DepEvents.size();
 
   void *alloc = nullptr;
 
   ur_event_handle_t Event = nullptr;
   // If a graph is present do the allocation from the graph memory pool instead.
   if (auto Graph = h.getCommandGraph(); Graph) {
-    alloc = Graph->getMemPool().malloc(size, kind);
+    // Get dependent graph nodes from any events
+    auto DepNodes = Graph->getNodesForEvents(DepEvents);
```

Comment:
There seems to be some  code duplication. Would it make sense to share some of the code between async_malloc and async_malloc_from_pool()?

Reply:
Moved this to a separate helper function